### PR TITLE
Sign-in and sign-up

### PR DESCRIPTION
**Pull request by YarikLysak**

### CODE REVIEW

File: src/app/auth/sign-up/sign-up.component.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  constructor(private fb: FormBuilder, private store: Store<AppState>, private validation: ValidationService) {}
+
+  ngOnInit() {
+    this.buildForm();
+    this.authErrors$.subscribe(errors => {
+      for (let err in errors) {
+        const control = this.signUpForm.controls[err];
+        control ? control.setErrors({ [err]: errors[err] }) : this.signUpForm.setErrors({ [err]: errors[err] });
```
Comment (dmitry-honcharuk) [inline src/app/auth/sign-up/sign-up.component.ts]: please rewrite to `if/else` statement, ternary operator without `assignment / return` is not a good practise

File: src/app/auth/sign-up/sign-up.component.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    });
+  }
+
+  constructor(private fb: FormBuilder, private store: Store<AppState>, private validation: ValidationService) {}
+
+  ngOnInit() {
+    this.buildForm();
+    this.authErrors$.subscribe(errors => {
+      for (let err in errors) {
+        const control = this.signUpForm.controls[err];
```
Comment (dmitry-honcharuk) [inline src/app/auth/sign-up/sign-up.component.ts]: should have `errors.hasOwnProperty(error)`

File: src/app/auth/store/auth.actions.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { createAction, props } from '@ngrx/store';
+import { SignUpData, SignInData } from '../shared/authData.model';
+
+export const login = createAction('[AUTH] Login', props<{ loginData: SignInData }>());
+export const register = createAction('[AUTH] Register', props<{ signUpData: SignUpData }>());
+
+export const authError = createAction('[AUTH] Register Errors', props<{ errors: any }>());
```
Comment (dmitry-honcharuk) [inline src/app/auth/store/auth.actions.ts]: `[AUTH] Register Errors` should be `Auth` instead if `Register`

File: src/app/auth/sign-up/sign-up.component.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import { register, logout } from '../store/auth.actions';
+
+import { emailValidator } from '../shared/validators/email.validator';
+import { passwordValidator } from '../shared/validators/password.validator';
+import { confirmPasswordValidator } from '../shared/validators/confirm-password.validator';
+import { ValidationService } from '../shared/validation.service';
+import { getAuthErrors } from '../store/auth.selector';
+
+@Component({
+  selector: 'sign-up',
```
Comment (dmitry-honcharuk) [inline src/app/auth/sign-up/sign-up.component.ts]: tslint considers this a warning:
`The selector should be prefixed by "app"`

Comment (YarikLysak) [reply src/app/auth/sign-up/sign-up.component.ts]: i have removed it from settings if it necessary i move it back

File: src/app/auth/shared/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    return this.http.post<string>(`api/auth/register`, signUpData);
+  }
+
+  login(loginData: SignInData): Observable<string> {
+    return this.http.post<string>(`api/auth/login`, loginData);
+  }
+
+  setToken(token: string): void {
+    localStorage.setItem('USER_TOKEN', token);
+    this.router.navigateByUrl('/home');
```
Comment (dmitry-honcharuk) [inline src/app/auth/shared/auth.service.ts]: why `setToken` method should do anything except setting the token?

File: src/app/auth/shared/validators/password.validator.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { FormControl } from '@angular/forms';
+
+export function passwordValidator(control: FormControl): null | object {
+  const regExp = /^(?=.*[0-9])[a-zA-Z0-9!@#$%^&*]{6,100}$/;
```
Comment (dmitry-honcharuk) [inline src/app/auth/shared/validators/password.validator.ts]: these constrains does not match with BE ones

**Comment by dmitry-honcharuk** [comment]
Email field doesn't show a "taken email" error on sign up form
image

**Comment by dmitry-honcharuk** [comment]
User should stay on home page after page reload
( currently user is being redirected back to sign-in page )

**Comment by dmitry-honcharuk** [comment]
User should not be able to register with _**Name**_  field filled with space
image

File: src/app/auth/shared/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  login(loginData: SignInData): Observable<string> {
+    return this.http.post<string>(`api/auth/login`, loginData);
+  }
+
+  setToken(token: string): void {
+    localStorage.setItem('USER_TOKEN', token);
+  }
+
+  ifHasToken(): boolean {
```
Comment (dmitry-honcharuk) [inline src/app/auth/shared/auth.service.ts]: simple `hasToken` is better naming

File: src/app/auth/store/auth.actions.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { createAction, props } from '@ngrx/store';
+import { SignUpData, SignInData } from '../shared/authData.model';
+
+export const login = createAction('[AUTH] Login', props<{ loginData: SignInData }>());
+export const register = createAction('[AUTH] Register', props<{ signUpData: SignUpData }>());
+export const ifHasToken = createAction('[AUTH] Auth If Has Token');
```
Comment (dmitry-honcharuk) [inline src/app/auth/store/auth.actions.ts]: `ifHasToken`  is not very good name for actions, it's not clear what this action is representing.
Looking on redux action naming convention, actions should have names that:

1. represent _**action**_ that happened (e.g.: COUNT_INCREASED) - _more preferred_
or
1. represent **_instruction_** to do some action  (e.g.: INCREASE_COUNT)  - _less preferred_

Comment (dmitry-honcharuk) [reply src/app/auth/store/auth.actions.ts]: For this particular situation you can name it something like `checkAuthentication` or whatever

File: src/app/auth/shared/validation.service.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+@Injectable({
+  providedIn: 'root'
+})
+export class ValidationService {
+  errorMessage({ errors, touched }): null | string {
+    for (const err in errors) {
+      if (errors.hasOwnProperty(err) && touched) {
+        const config = {
+          required: 'Required',
+          email: 'Invalid email address!',
```
Comment (dmitry-honcharuk) [inline src/app/auth/shared/validation.service.ts]: Please remove exclamation mark, you shouldn't yell at the user  :)

File: src/app/auth/shared/validators/name.validator.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+import { FormControl } from '@angular/forms';
+
+export function nameValidator(control: FormControl): null | object {
+  return control.value.trim() ? null : { invalidName: `Enter correct name.` };
```
Comment (dmitry-honcharuk) [inline src/app/auth/shared/validators/name.validator.ts]: Please change message to `Name cannot be empty.` (also you don't need template string here , you can use single quotes `'`)

File: src/app/auth/store/auth.reducer.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  errors: null
+};
+
+const reducer = createReducer(
+  initialState,
+  on(authSuccess, (state, { isAuth }) => {
+    return { ...state, isAuth, errors: null };
+  }),
+  on(logoutSuccess, (state, { isAuth }) => {
+    return { ...state, isAuth, errors: null };
```
Comment (dmitry-honcharuk) [inline src/app/auth/store/auth.reducer.ts]: I can't really see any reason to path any payload along with `logoutSuccess` action, its name represents everything you need, you can just set `isAuth: false` on this action

**Pull request merged**